arch/arm/src/cxd56xx/cxd56_irq.c

```diff
@@ -115,17 +112,17 @@ static uint64_t g_intstack_alloc[INTSTACK_ALLOC >> 3];
 
 const uint32_t g_cpu_intstack_top[CONFIG_SMP_NCPUS] =
 {
-  (uint32_t)g_intstack_alloc + INTSTACK_SIZE,
+  (uint32_t)g_intstack_alloc + INTSTACK_SIZE - 8,
```

Comment:
Is -8 really right? AAPCS require the full descending stack:
https://developer.arm.com/documentation/dui0204/j/writing-arm-assembly-language/load-and-store-multiple-register-instructions/implementing-stacks-with-ldm-and-stm
If My understand is correct, all -8/-4 should be removed instead. Other arch has the similar problem too.

Reply:
![image](https://user-images.githubusercontent.com/18066964/96555406-51dcbc80-12ea-11eb-8b48-195dd9721903.png)
![image](https://user-images.githubusercontent.com/18066964/96555504-75076c00-12ea-11eb-90b3-56a22b4e9431.png)
![image](https://user-images.githubusercontent.com/18066964/96555642-9c5e3900-12ea-11eb-8dac-9170706e0378.png)

Reply:
> so to avoid the similar issue happen again, it's better to fix the problem from root by removing -8 at all.

@xiaoxiang781216 
Thanks for your comments.
I thought the current implementation of SMP_STACK_TOP was correct.
So do you mean I should fix SMP_STACK_TOP as well ?

Reply:
I think that adjustment is necessary to get the address of the last element in the stack, at least for the case of how g_intstack_alloc is allocated (a 64bit array)

Reply:
ARM(and many other arch) AAPCS require that push implement like this:
```
sub sp, #4/#8
str val, [sp]
```
so the pointer should point to after the end of stack. -8 make the code complexity and waste a little memory, or is there a special reason to do in this way, @patacongo? I have worked with many RTOS and bootloader, but never see the type of special code.

Reply:
>so the pointer should point to after the end of stack. -8 make the code complexity and waste a little memory, or is >there a special reason to do in this way, @patacongo? I have worked with many RTOS and bootloader, but never see >the type of special code.

@xiaoxiang781216 
Thanks for your comments.

The situation is the same for non-SMP case as well.
The following code shows arch/arm/src/armv7-a/arm_vectors.S
but similar code exists for other ARM architectures.

As you can see and yes as you pointed out,
$sp is not subtracted in setirqstack then save a value to the stack.

```
...
if CONFIG_ARCH_INTERRUPTSTACK > 7                                                                                                                                          
#ifndef CONFIG_SMP                                                                                                                                                          
.Lirqstackbase:                                                                                                                                                             
    .word   g_intstackbase                                                                                                                                                  
#endif
...
if !defined(CONFIG_SMP) && CONFIG_ARCH_INTERRUPTSTACK > 7                                                                                                                  
    .macro  setirqstack, tmp1, tmp2                                                                                                                                         
    ldr     sp, .Lirqstackbase      /* SP = IRQ stack top */                                                                                                                
    .endm                                                                                                                                                                   
#endif 
...
    /* Call arm_decodeirq() on the interrupt stack */                                                                                                                       
                                                                                                                                                                            
    setirqstack r1, r3              /* SP = IRQ stack top */                                                                                                              
    str     r0, [sp]                /* Save the user stack pointer */                                                                                                       
    mov     r4, sp                  /* Save the SP in a preserved register */                                                                                               
    bic     sp, sp, #7              /* Force 8-byte alignment */                                                                                                            
    bl      arm_decodeirq           /* Call the handler */  
...
```

Rather, in this caes g_intstackbase is subtracted explicitly (i.e. -4 but I think this should be -8) here for non-SMP case.

```
...
g_intstackalloc:                                                                                                                                                            
    .skip   ((CONFIG_ARCH_INTERRUPTSTACK & ~7) - 4)                                                                                                                         
g_intstackbase:                                                                                                                                                             
    .skip   4                                                                                                                                                               
    .size   g_intstackbase, 4     
...             
```


I agree that this kind of code is complicated, so should be more simple.
However, I think this should done as another task to refactor the code,
because the impact would be bigger than the current PR.

What do you think?

Reply:
Sure, it's better to address all -8/-4 issue in another patch.

Reply:
>Sure, it's better to address all -8/-4 issue in another patch.

@xiaoxiang781216 
Thanks for accepting my proposal.
I will create another PR for that issue. (Perhaps, firstly for ARM)

Now, this PR conforms to existing rules.
Can you accept this PR?

Reply:
Yes, let me merge it.